cassiopeia/tools/branch_length_estimator/IIDExponentialMLE.py

```diff
@@ -0,0 +1,174 @@
+"""
+This file stores a subclass of BranchLengthEstimator, the IIDExponentialMLE.
+Briefly, this model assumes that CRISPR/Cas9 mutates each site independently
+and identically, with an exponential waiting time.
+"""
+import cvxpy as cp
+import numpy as np
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.mixins import IIDExponentialMLEError
+
+from .BranchLengthEstimator import BranchLengthEstimator
+
+
+class IIDExponentialMLE(BranchLengthEstimator):
+    """
+    MLE under a model of IID memoryless CRISPR/Cas9 mutations.
+
+    This model assumes that CRISPR/Cas9 mutates each site independently and
+    identically, with an exponential waiting time. The MLE under this model
+    is a special kind of convex optimization problem known as an exponential
+    cone program, which can be readily solved with off-the-shelf (open source)
+    solvers.
+
+    This estimator requires that the ancestral characters be provided (these
+    can be imputed with CassiopeiaTree's reconstruct_ancestral_characters
+    method if they are not known, which is usually the case for real data).
+
+    Because branch lengths and CRISPR/Cas9 mutation rate are not identifiable,
+    this estimator assumes that the tree has a depth of 1. In other words,
+    the estimated tree will have depth 1. The estimated mutation rate under
+    this unit-depth assumption will be stored as an attribute called
+    `mutation_rate`.
+
+    Missing states are treated as missing at random by the model.
+
+    Args:
+        minimum_branch_length: Estimated branch lengths will be constrained to
+            have length at least this value. By default it is set to 0.01,
+            since the MLE tends to collapse mutationless edges to length 0.
+        solver: Convex optimization solver to use. Can be "SCS", "ECOS", or
+            "MOSEK". Note that "MOSEK" solver should be installed separately.
+        verbose: Verbosity level.
+
+    Attributes:
+        log_likelihood: The log-likelihood of the training data under the
+            estimated model.
+        mutation_rate: The estimated CRISPR/Cas9 mutation rate, assuming that
+            the tree has depth exactly 1.
+    """
+
+    def __init__(
+        self,
+        minimum_branch_length: float = 0.01,
+        verbose: bool = False,
+        solver: str = "SCS",
+    ):
+        allowed_solvers = ["ECOS", "SCS", "MOSEK"]
+        if solver not in allowed_solvers:
+            raise ValueError(
+                f"Solver {solver} not allowed. Allowed solvers: {allowed_solvers}"
+            )  # pragma: no cover
+        self._minimum_branch_length = minimum_branch_length
+        self._verbose = verbose
+        self._solver = solver
+
+    def estimate_branch_lengths(self, tree: CassiopeiaTree) -> None:
+        r"""
```

Comment:
Our code best-practices are to start with a single-line (less than 80 characters) docstring header.